frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -1305,6 +1305,114 @@ const IR::Node* TypeInference::postorder(IR::Concat* expression) {
+*/
+const IR::Node* TypeInference::postorder(IR::EntriesList* el) {
+    LOG1("TypeInference: visiting: " << el);
+    if (done()) return el;
+    auto table = findContext<IR::P4Table>();
+    BUG_CHECK(table != nullptr, "%1% entries not within a table", el);
+    const IR::Key* key = table->getKey();
+    if (key == nullptr) {
+        ::error("Could not find key for table %1%", table);
+        return el;
+    }
+
+    // the key must have been typechecked already.
+    if (!typeMap->contains(key)) {
+        ::error("Table initializer entries must occur after key definitions (table %1%)", table);
+        return el;
+    }
+    return el;
+}
+/**
+    typecheck a table initializer entry
+
+    The invariants are:
+    - table keys and entry keys must have the same length
+    - entry key elements must be compile time constants
+    - actionRefs in entries must be in the action list
+
+    Moreover, the EntriesList visitor should have checked for the table
+    invariants, and thus, those checks are now bug checks.
+*/
+const IR::Node* TypeInference::postorder(IR::Entry* entry) {
+    LOG1("TypeInference: visiting: " << entry);
+    if (done()) return entry;
+    auto table = findContext<IR::P4Table>();
+    BUG_CHECK(table != nullptr, "%1% entries not within a table", entry);
+    const IR::Key* key = table->getKey();
+    BUG_CHECK(key != nullptr, "table %1% has no keys", table);
+    auto keyTuple = getType(key);
+    BUG_CHECK(keyTuple != nullptr, "table %1% has no types for keys", table);
+    auto entryKeyType = getType(entry->keys);
+    if (entryKeyType == nullptr)
+        return entry;
+    if (entryKeyType->is<IR::Type_Set>())
+        entryKeyType = entryKeyType->to<IR::Type_Set>()->elementType;
+
+    auto tvs = unify(entry, keyTuple, entryKeyType, true);  // TypeVariableSubstitution
+    if (tvs == nullptr)
+        return entry;
+    ConstantTypeSubstitution cts(tvs, typeMap);
+    auto ks = cts.convert(entry->keys);
+
+    for (auto ke : *ks->to<IR::ListExpression>()->components)
+        if (!isCompileTimeConstant(ke))
+            ::error("Key entry must be a compile time constant: %1%", ke);
+
+    if (ks != entry->keys)
+        entry = new IR::Entry(entry->srcInfo, entry->annotations,
+                              ks->to<IR::ListExpression>(), entry->action);
+
+    auto actionRef = entry->getAction();
+    if (!actionRef->is<IR::MethodCallExpression>())
+     ::error("Invalid action %1% in entry");
+    auto actionCall = actionRef->to<IR::MethodCallExpression>();
+    auto actionName = actionCall->method->to<IR::PathExpression>()->path->name;
+    bool found = false;
+    for (auto ale : *table->getActionList()->actionList) {
+     auto expr = ale->expression;
+     if (expr->is<IR::MethodCallExpression>())
+         expr = expr->to<IR::MethodCallExpression>()->method;
+     if (expr->is<IR::PathExpression>() &&
+         expr->to<IR::PathExpression>()->path->name == actionName) {
```

Comment:
I don't know if this test is enough; you could have a global action with the same name.
What you should really do is to use the referenceMap to make sure both PathExpressions refer to the same declaration.